lib/datadog/di/probe_manager.rb

```diff
@@ -0,0 +1,182 @@
+
+          installed_probes[probe.id] = probe
+          payload = probe_notification_builder.build_installed(probe)
+          probe_notifier_worker.add_status(payload)
+          # The probe would only be in the pending probes list if it was
+          # previously attempted to be installed and the target was not loaded.
+          # Always remove from pending list here because it makes the
+          # API smaller and shouldn't cause any actual problems.
+          pending_probes.delete(probe.id)
+          true
+        rescue Error::DITargetNotDefined => exc
+          pending_probes[probe.id] = probe
+          false
+        end
+      rescue => exc
+        # In "propagate all exceptions" mode we will try to instrument again.
+        raise if settings.dynamic_instrumentation.internal.propagate_all_exceptions
+
+        logger.warn("Error processing probe configuration: #{exc.class}: #{exc}")
+        telemetry&.report(exc, description: "Error processing probe configuration")
+        # TODO report probe as failed to agent since we won't attempt to
+        # install it again.
+
+        # TODO add top stack frame to message
+        failed_probes[probe.id] = "#{exc.class}: #{exc}"
+
+        raise
+      end
+
+      def remove_other_probes(probe_ids)
+        pending_probes.values.each do |probe|
+          unless probe_ids.include?(probe.id)
+            pending_probes.delete(probe.id)
+          end
+        end
+        installed_probes.values.each do |probe|
+          unless probe_ids.include?(probe.id)
+            begin
+              instrumenter.unhook(probe)
+              # Only remove the probe from installed list if it was
+              # successfully de-instrumented. Active probes do incur overhead
+              # for the running application, and if the error is ephemeral
+              # we want to try removing the probe again at the next opportunity.
+              #
+              # TODO give up after some time?
+              installed_probes.delete(probe.id)
+            rescue => exc
+              raise if settings.dynamic_instrumentation.internal.propagate_all_exceptions
+              # Silence all exceptions?
+              # TODO should we propagate here and rescue upstream?
+              logger.warn("Error removing probe #{probe.id}: #{exc.class}: #{exc}")
+              telemetry&.report(exc, description: "Error removing probe #{probe.id}")
+            end
+          end
+        end
+      end
+
+      private def install_pending_method_probes(cls)
+        # TODO search more efficiently than linearly
+        pending_probes.each do |probe_id, probe|
+          if probe.method?
+            # TODO move this stringification elsewhere
+            if probe.type_name == cls.name
+              # TODO is it OK to hook from trace point handler?
+              # TODO the class is now defined, but can hooking still fail?
+              # TODO pass rate_limiter here, need to get it from somewhere
```

Comment:
Isn't the rate limiter available here, in `probe.rate_limiter`?

Reply:
Yes, deleted the comment.